x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/CreateAndFollowIndexAction.java

```diff
@@ -0,0 +1,309 @@
+                    createFollowIndex(leaderIndexMetadata, request, listener);
+                }, listener::onFailure));
+            }
+        }
+
+        private void createFollowIndex(IndexMetaData leaderIndexMetaData, Request request, ActionListener<Response> listener) {
+            if (leaderIndexMetaData == null) {
+                listener.onFailure(new IllegalArgumentException("leader index [" + request.getFollowRequest().getLeaderIndex() +
+                    "] does not exist"));
+                return;
+            }
+
+            ActionListener<Boolean> handler = ActionListener.wrap(
+                result -> {
+                    if (result) {
+                        initiateFollowing(request, listener);
+                    } else {
+                        listener.onResponse(new Response(true, false, false));
+                    }
+                },
+                listener::onFailure);
+            clusterService.submitStateUpdateTask("follow_index_action", new AckedClusterStateUpdateTask<Boolean>(request, handler) {
+
+                @Override
+                protected Boolean newResponse(boolean acknowledged) {
+                    return acknowledged;
+                }
+
+                @Override
+                public ClusterState execute(ClusterState currentState) throws Exception {
+                    IndexMetaData currentIndex = currentState.metaData().index(request.getFollowRequest().getFollowIndex());
+                    if (currentIndex != null) {
+                        throw new ResourceAlreadyExistsException(currentIndex.getIndex());
+                    }
+
+                    MetaData.Builder mdBuilder = MetaData.builder(currentState.metaData());
+                    if (request.getFollowRequest().getFollowIndex() != null) {
+                        IndexMetaData.Builder imdBuilder = IndexMetaData.builder(leaderIndexMetaData);
+
+                        Settings.Builder settingsBuilder = Settings.builder();
+                        settingsBuilder.put(leaderIndexMetaData.getSettings());
+                        // TODO: do we want leader and follow index to have the same UUID?
+                        // (Overwriting UUID here, because otherwise we can't follow indices in the same cluster)
+                        settingsBuilder.put(IndexMetaData.SETTING_INDEX_UUID, UUIDs.randomBase64UUID());
+                        settingsBuilder.put(IndexMetaData.SETTING_INDEX_PROVIDED_NAME, request.getFollowRequest().getFollowIndex());
+                        settingsBuilder.put(CcrSettings.CCR_FOLLOWING_INDEX_SETTING.getKey(), true);
+
+                        imdBuilder.settings(settingsBuilder);
+                        imdBuilder.index(request.getFollowRequest().getFollowIndex());
+                        for (int shardId = 0 ; shardId < leaderIndexMetaData.getNumberOfShards(); shardId++) {
+                            imdBuilder.putInSyncAllocationIds(shardId, new HashSet<>());
```

Comment:
I don't think this is strictly needed (it should be the default; see [`IndexMetaData#build`](https://github.com/elastic/elasticsearch/blob/916bf9d26d6292dcb9b229123c58a86f0e741ffd/server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java#L1036-L1044)). Did you find otherwise?

Reply:
Yes, I did. Shards of the follow index would never get started (allocation status `no_valid_shard_copy`).

Reply:
Okay; this is happening because we copy all of the leader index metadata (which brings along with it the in-sync shard copies on the leader). Then, we have to overwrite with an empty in-sync shard set. I don't think we should do this, I lean towards thinking that we should only copy explicitly what we need from the leader index rather than copying everything and overwriting.

Reply:
> I lean towards thinking that we should only copy explicitly what we need from the leader index rather than copying everything and overwriting.

+1 , at least at the current stage where we focus on data and not cluster level metadata (with the exception of mappings).